climakitae/threshold_tools.py

```diff
@@ -0,0 +1,613 @@
+import numpy as np
+import pandas as pd
+import xarray as xr
+from scipy import stats
+
+import lmoments3 as lm
+from lmoments3 import distr as ldistr
+from lmoments3 import stats as lstats
+
+import matplotlib.pyplot as plt
+from matplotlib import cm
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
+import geoviews as gv
+
+import holoviews as hv
+from holoviews import opts
+import hvplot.pandas
+import hvplot.xarray
+
+from .visualize import get_frequency_plot, get_geospatial_plot
+
+#####################################################################
+
+
+def get_ams(da, extremes_type="max"):
+
+    extremes_types = ["max"]
+    if extremes_type not in extremes_types:
+        raise ValueError(
+            "invalid extremes type. expected one of the following: %s" % extremes_types
+        )
+
+    if extremes_type == "max":
+        ams = da.resample(time="A").max(keep_attrs=True)
+        ams.attrs["extreme value extraction method"] = "block maxima"
+        ams.attrs["extremes type"] = "maxima"
+        ams.attrs["block size"] = "1 year"
+        ams.attrs["timeseries type"] = "annual max series"
+
+    # if extremes_type == 'min':
+    #     ams = da.resample(time='A').min(keep_attrs=True)
+    #     ams.attrs['extreme value extraction method'] = 'block maxima'
+    #     ams.attrs['extremes type'] = 'minima'
+    #     ams.attrs['block size'] = '1 year'
+    #     ams.attrs['timeseries type'] = 'annual min series'
+
+    return ams
+
+
+#####################################################################
+
+
+def get_lmoments(ams, distr="gev", multiple_points=True):
+
+    distrs = ["gev", "gumbel", "weibull", "pearson3", "genpareto"]
+    if distr not in distrs:
+        raise ValueError(
+            "invalid distr type. expected one of the following: %s" % distrs
+        )
+
+    ams_attributes = ams.attrs
+
+    if multiple_points == True:
+        ams = ams.stack(allpoints=["x", "y"]).squeeze().groupby("allpoints")
```

Comment:
It doesn't look like is needed since `unstack()` just undoes this below. It's always faster to reduce the overall number of dask tasks.

Reply:
yes, I'm not seeing what @bkg  is getting at -- you need to stack and unstack to do this operation over all points if the data in question is an xarray object with multiple grid-cells (maybe instead of "multiple points" it'd be clearer to say if "spatial grid"?)